File: ConsoleApp1/InputScanner.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace ConsoleApp1
{
    public class InputScanner:IInputScanner
    {
        private readonly InvertedIndex _index;

        public InputScanner(InvertedIndex index)
        {
            this._index = index;
        }


        public IEnumerable<string> GetOrder(string input)
        {
            var inputSplit = SplitInput(input);
            var plusStrings = new List<string>();
            var minusStrings = new List<string>();
            var normalStrings = new List<string>();
            foreach (var s in inputSplit)
            {
                addItemToOneOfThreeArrayLists(s, plusStrings, minusStrings, normalStrings);
            }

            return Processes(_index, plusStrings, minusStrings, normalStrings);
        }

        private static IEnumerable<string> SplitInput(string input)
        {
            return input.Split("(\\s+)");
        }


        private static HashSet<string> Processes(InvertedIndex index, List<string> plusStrings,
            List<string> minusStrings,
            IEnumerable<string> normalStrings)
        {
            var answer = index.Search(plusStrings);

            var toDelete = index.Search(minusStrings);
            var commons = normalStrings.Select(normalString => new List<string> {normalString}).Select(index.Search)
                .ToList();

            answer = index.FindCommonFiles(answer, commons);
            answer = index.DeleteGivenFiles(answer, toDelete);
            return answer;
        }

        private static void addItemToOneOfThreeArrayLists(string iString, ICollection<string> plusStrings,
            ICollection<string> minusStrings,
            ICollection<string> normalStrings)
        {
            var pattern = new Regex("^\\+(.+)$");
            var matcher = pattern.Match(iString);
            var patternTwo = new Regex("^-(.+)$");
            var matcherTwo = patternTwo.Matches(iString);
            if (matcher.Success)
            {
                var toAdd = matcher.Value;
                plusStrings.Add(toAdd);
            }
            else if (matcherTwo.Count > 0)
            {
                var toAdd = matcherTwo[0].Value;
                minusStrings.Add(toAdd);
            }
            else normalStrings.Add(iString);
        }
    }
}

File: src/main/java/InputScanner.java
import java.util.ArrayList;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class InputScanner {
    private InvertedIndex index;

    public InputScanner(InvertedIndex index) {
        this.index = index;
    }


    public Set<String> getOrder(String input) {
        String[] inputSplit = splitInput(input);
        ArrayList<String> plusStrings = new ArrayList<>();
        ArrayList<String> minusStrings = new ArrayList<>();
        ArrayList<String> normalStrings = new ArrayList<>();
        for (String string : inputSplit)
            addItemToOneOfThreeArrayLists(string, plusStrings, minusStrings, normalStrings);
        return processes(index, plusStrings, minusStrings, normalStrings);
    }

    private String[] splitInput(String input) {
        return input.split("(\\s+)");
    }


    private Set<String> processes(InvertedIndex index, ArrayList<String> plusStrings, ArrayList<String> minusStrings, ArrayList<String> normalStrings) {
        Set<String> answer = index.search(plusStrings);
        Set<String> toDelete = index.search(minusStrings);
        ArrayList<Set<String>> commons = new ArrayList<>();
        for (String normalString : normalStrings) {
            ArrayList<String> arrayList = new ArrayList<>();
            arrayList.add(normalString);
            commons.add(index.search(arrayList));
        }
        answer = index.findCommonFiles(answer, commons);
        answer = index.deleteGivenFiles(answer, toDelete);
        return answer;
    }

    private void addItemToOneOfThreeArrayLists(String string, ArrayList<String> plusStrings, ArrayList<String> minusStrings, ArrayList<String> normalStrings) {
        Pattern patternForPlusWords = Pattern.compile("^\\+(.+)$");
        Matcher matcherForPlusWords = patternForPlusWords.matcher(string);
        Pattern patternForMinusWords = Pattern.compile("^-(.+)$");
        Matcher matcherForMinusWords = patternForMinusWords.matcher(string);
        if (matcherForPlusWords.find()) {
            String a = matcherForPlusWords.group(1);
            plusStrings.add(a);
        } else if (matcherForMinusWords.find()) {
            String a = matcherForMinusWords.group(1);
            minusStrings.add(a);
        } else normalStrings.add(string);
    }

}
